### PR TITLE
Fix #19445: Text not cleared using Ctrl+Backspace

### LINKED ISSUE #19445: Ctrl+backspace visual glitch
### Operating System

Windows 10, 64-bit

### OpenRCT2 build

OpenRCT2, v0.4.3-ef25274

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This bug is a graphical glitch or error

### Describe the issue

When clearing a ride name with ctrl-backspace, the old name is not properly cleared from the input widget. This only started happening in the last few days.

### Steps to reproduce

1. Open the ride window for any attraction.
2. Click the rename button.
3. Type ctrl-backspace.
4. The cursor moves to the front of the textbox, but the old name is still visible.
5. Type a new name. It will overwrite the old name one character at a time.
6. Click OK to save.
7. None of the old name is visible.

So the _behavior_ of the textbox is still valid. But the _display_ is kinda messed up. A video of this would be helpful but I'm not at my PC right now

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by ZehMatt**
Unfortunately there was no easy fix for this, the issue was caused by the TextComposition only writing a null terminator at the first offset to clear it but later would use string_view with its actual size. I refactored the code to use u8string to be always the buffer and TextComposition will now directly work with that object rather than having only the pointer and separate size. The maximum text limit is also now considering codepoints instead of bytes which was previously the other way.

Closes #19445

### CODE REVIEW

File: src/openrct2-ui/TextComposition.cpp [hunk 1]
```diff
@@ -36,19 +36,16 @@ bool TextComposition::IsActive()
     return SDL_IsTextInputActive() && _session.Buffer != nullptr;
 }
 
-TextInputSession* TextComposition::Start(utf8* buffer, size_t bufferSize)
+TextInputSession* TextComposition::Start(u8string& buffer, size_t maxLength)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/TextComposition.cpp:39]: Can't we just own the buffer in TextComposition

File: distribution/changelog.txt [hunk 2]
```diff
@@ -48,6 +48,7 @@
 - Fix: [#19379] "No platforms" station style shows platforms on the Junior Roller Coaster.
 - Fix: [#19380] Startup crash when no sequences are installed and random sequences are enabled.
 - Fix: [#19391] String corruption caused by an improper buffer handling in GfxWrapString.
+- Fix: [#19445] Text not cleared using Ctrl+Backspace.
```
Comment (duncanspumpkin) [inline distribution/changelog.txt:51]: Hasn't this bug only just been added. I.e. no changelog required.

Comment (ZehMatt) [reply distribution/changelog.txt:51]: You are right.

File: src/openrct2-ui/TextComposition.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2-ui/TextComposition.cpp:39]: There is currently a global shared among Widget.cpp and Window.cpp but probably.

**Comment by ZehMatt** [comment]
Deleting special characters seem to behave a bit weird currently, gonna put this into draft until I figured that out.

File: src/openrct2-ui/TextComposition.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2-ui/TextComposition.cpp:39]: Unfortunately no, unless each input receives its own TextInputSession instance which is currently not the case, can try to make this cleaner after the fix.

**Comment by ZehMatt** [comment]
I've implemented #19468 in here as well as it affected the same lines.

This works the same way as in Notepad++

**Comment by Broxzier** [comment]
Looks nice in the recording already. It now checks whitespace, but perhaps it needs to check word bounds instead. How does it deal with names like "Batman: Arkham Asylum"?

In Notepad++ it goes like:
 1. `Batman: Arkham Asylum`
 2. `Batman: Arkham `
 3. `Batman: `
 4. `Batman`
 5. ``

In Firefox is goes like:
 1. `Batman: Arkham Asylum`
 7. `Batman: Arkham `
 8. `Batman: `
 9. ``

**Review by Broxzier — APPROVED** [review]

**Comment by Broxzier** [comment]
I guess a changelog entry for the change in behaviour is warranted.

**Pull request merged**